src/ck-ldb/CentralLB.C

```diff
@@ -883,29 +876,26 @@ void CentralLB::removeNonMigratable(LDStats* stats, int count)
 
   // check if we have non-migratable objects
   int have = 0;
-  for (i=0; i<stats->n_objs; i++) 
+  for (const auto& odata : stats->objData)
   {
-    LDObjData &odata = stats->objData[i];
     if (!odata.migratable) {
       have = 1; break;
     }
   }
   if (have == 0) return;
 
-  CkVec<LDObjData> nonmig;
-  CkVec<int> new_from_proc, new_to_proc;
-  nonmig.resize(stats->n_migrateobjs);
-  new_from_proc.resize(stats->n_migrateobjs);
-  new_to_proc.resize(stats->n_migrateobjs);
-  int n_objs = 0;
-  for (i=0; i<stats->n_objs; i++) 
+  std::vector<LDObjData> nonmig;
+  std::vector<int> new_from_proc, new_to_proc;
+  nonmig.reserve(stats->n_migrateobjs);
+  new_from_proc.reserve(stats->n_migrateobjs);
+  new_to_proc.reserve(stats->n_migrateobjs);
+  for (i=0; i<stats->objData.size(); i++)
   {
     LDObjData &odata = stats->objData[i];
     if (odata.migratable) {
-      nonmig[n_objs] = odata;
-      new_from_proc[n_objs] = stats->from_proc[i];
-      new_to_proc[n_objs] = stats->to_proc[i];
-      n_objs ++;
+      nonmig.push_back(odata);
```

Comment:
I think this is better as emplace_back?

Reply:
Since this is adding an object that already exists, I don't think there would be any difference between `emplace_back` and `push_back` in this case. I'd prefer to keep it as `push_back` to make it clear that it's going to be calling a copy constructor (see [this](https://abseil.io/tips/112) for more rationale).

Reply:
I think push_back will result in calling the copy constructor to create a new object, which will then be moved into the vector, whereas emplace_back will call the copy constructor in place, avoiding the move into the vector. The extent of the performance difference depends on LDObjData, its size, whether it has a move constructor that is different from its copy constructor.

Reply:
I wrote a test program to check this, and both `push_back` and `emplace_back` just call a single copy constructor per invocation, directly constructing into the target location in the vector in the case where the argument is of their template type `T`. Behavior is only different in the case where it can use perfect forwarding to avoid having to do a premature construction when arguments to a constructor are passed in, in which case it can defer calling the constructor until it can be done in place.

For example, as [seen in libc++](https://github.com/llvm/llvm-project/blob/master/libcxx/include/vector), both `push_back` and `emplace_back` are essentially calls to `__construct_one_at_end` with arguments passed in in different ways (either directly as a `const T&` or with `std::move` for `push_back` and via `std::forward` for `emplace_back`). In the case of `nonmig.push_back(odata)`, `odata` gets passed as a `const_reference` until `__construct_one_at_end`, where it is then used for a copy construction. For `nonmig.emplace_back(odata)`, `odata` gets passed as a perfectly forwarded argument until `__construct_one_at_end`, where it is then used for a copy construction in the same way. In both cases, there is a single copy construction.

If you want to avoid calling the copy constructor completely when adding an existing object `T obj` to a `vector<T>`, you can use either `push_back(std::move(obj))` or `emplace_back(std::move(obj))`, both are identical.